tools/pubsys/src/aws/ami/register/mk_amispec.rs

```diff
@@ -0,0 +1,833 @@
+    ) -> HashMap<String, BlockDeviceRenderContext> {
+        let (os_volume_size, data_volume_size) = image_layout.publish_image_sizes_gib();
+
+        let mut render_ctx = HashMap::new();
+        render_ctx.insert(
+            "root".into(),
+            BlockDeviceRenderContext {
+                device_name: ROOT_DEVICE_NAME.into(),
+                snapshot_id: self.os_snapshot.clone(),
+                volume_size: os_volume_size as u64,
+            },
+        );
+
+        if let Some(data_snapshot) = &self.data_snapshot {
+            render_ctx.insert(
+                "data".into(),
+                BlockDeviceRenderContext {
+                    device_name: DATA_DEVICE_NAME.into(),
+                    snapshot_id: data_snapshot.clone(),
+                    volume_size: data_volume_size as u64,
+                },
+            );
+        }
+
+        render_ctx
+    }
+}
+
+#[derive(Debug, Snafu)]
+#[snafu(module)]
+pub(crate) enum AmispecError {
+    #[snafu(display("Attempted to configure EBS volume with invalid size: {}", source))]
+    InvalidEbsVolumeSize { source: InvalidValueError },
+
+    #[snafu(display("Could not find image layout for '{}'", variant_manifest.display_path()))]
+    MissingImageLayout {
+        variant_manifest: Box<VariantManifest>,
+    },
+
+    #[snafu(display("Failed to parse AMI spec template: {}", source))]
+    ParseAmiSpecTemplate { source: toml::de::Error },
+
+    #[snafu(display("Failed to render amispec for RegisterImages call: {}", source))]
+    RenderAmiSpec { source: amispec::TemplatedError },
+}
+
+#[cfg(test)]
+mod test {
+    use super::*;
+    use crate::aws::ami::register::AmiArgs;
+    use crate::aws::ami::VariantManifest;
+    use amispec::{
+        ebs, Architecture, BlockDeviceMapping, BootMode, SriovNetSupport, VirtualizationType,
+    };
+    use buildsys::manifest::ManifestInfo;
+    use maplit::hashmap;
+    use std::path::PathBuf;
+
+    use test_case::test_case;
+
+    const ROOT_SNAP_ID: &str = "os-snapshot";
+    const DATA_SNAP_ID: &str = "data-snapshot";
+    const DEFAULT_ROOT_VOLUME_SIZE: u64 = 2;
+    const DEFAULT_DATA_VOLUME_SIZE: u64 = 20;
+
+    lazy_static::lazy_static! {
+        static ref SPLIT_LAYOUT_SNAPSHOTS: BottlerocketSnapshots = BottlerocketSnapshots {
+            os_snapshot: ROOT_SNAP_ID.into(),
+            data_snapshot: Some(DATA_SNAP_ID.into()),
+        };
+
+        static ref UNIFIED_LAYOUT_SNAPSHOTS: BottlerocketSnapshots = BottlerocketSnapshots {
+            os_snapshot: ROOT_SNAP_ID.into(),
+            data_snapshot: None,
+        };
+    }
+
+    #[test_case(
+        AmiArgs {
+            os_image: PathBuf::new(),
+            data_image: None,
+            variant_manifest: VariantManifest::new(
+                toml::toml! {
+                    [package]
+                    name = "test-variant"
+
+                    [package.metadata.build-variant.image-layout]
+                    os-image-size-gib = 100
```

Comment:
With `partition-plan` unspecified, this should default to the "split" layout, so I'm surprised that it ended up with "unified".

Reply:
The data flow in `pubsys` is a bit hard to follow. It needs information about the variant being published to register correctly; however, `pubsys'` CLI arguments are in many cases redundant with the manifest. So now `pubsys` has to decide which to respect.

Both before and after this change, `pubsys` uses its CLI arguments to determine properties of the variant (only falling back to using the manifest itself when it can't, like when trying to determine what to do about Secure Boot.)

I'd really like to refactor it so that the decision of how to derive this data is more clear from the way that the software is written, rather than implied by reading some of the implementation details. It's not clear how long that would take me though, so I've stopped short of doing that here.